File: Math & Geometry/202. Happy Number.py
### Solution1 ###
class Solution:
    def isHappy(self, n: int) -> bool:
        visit = set()
        
        while n not in visit:
            visit.add(n)
            n = self.SquareSum(n)
            
            if n==1:
                return True
        return False
        
    def SquareSum(self, n: int) -> int:
        res = 0
        while n:
            res += (n%10) ** 2
            n //=10
        return res


### Solution2 ###
class Solution:
    def isHappy(self, n: int) -> bool:
        slow, fast = n, self.SquareSum(n)
        while slow != fast:
            fast = self.SquareSum(fast)
            fast = self.SquareSum(fast)
            slow = self.SquareSum(slow)
        
        return True if fast==1 else False
    
        
    def SquareSum(self, n: int) -> int:
        res = 0
        while n:
            res += (n%10) ** 2
            n //=10
        return res

File: 17.Math & Geometry/202. Happy Number.cpp
// Solution1 
class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> visit;
        
        while (visit.find(n) == visit.end()) {
            visit.insert(n);
            n = squareSum(n);
            if (n == 1) {
                return true;
            }
        }
        return false;
    }
    
private:
    int squareSum(int n) {
        int res = 0;
        while (n > 0) {
            res += pow((n%10), 2);
            n /= 10;
        }
        return res;
    }
};


// Solution2
class Solution {
public:
    bool isHappy(int n) {
        int slow = n;
        int fast = squareSum(n);
        
        while (slow != fast) {
            fast = squareSum(squareSum(fast));
            slow = squareSum(slow);
        }
        if (fast==1) {
            return true;
        } else {
            return false;
        }
    }
    
private:
    int squareSum(int n) {
        int res = 0;
        while (n > 0) {
            res += pow((n%10), 2);
            n /= 10;
        }
        return res;
    }
};


